sklearn/neural_network/multilayer_perceptron.py

```diff
@@ -0,0 +1,1248 @@
+            last, n_samples, activations, deltas, coef_grads, intercept_grads)
+
+        # Iterate over the hidden layers
+        for i in range(self.n_layers_ - 2, 0, -1):
+            deltas[i - 1] = safe_sparse_dot(deltas[i],
+                                            self.coefs_[i].T)
+            derivative = DERIVATIVES[self.activation]
+            deltas[i - 1] *= derivative(activations[i])
+
+            coef_grads, \
+                intercept_grads = self._compute_loss_grad(i - 1,
+                                                          n_samples,
+                                                          activations,
+                                                          deltas,
+                                                          coef_grads,
+                                                          intercept_grads)
+
+        return loss, coef_grads, intercept_grads
+
+    def _initialize(self, y, layer_units):
+        # set all attributes, allocate weights etc for first call
+        # Initialize parameters
+        self.n_iter_ = 0
+        self.t_ = 0
+        self.n_outputs_ = y.shape[1]
+
+        # Compute the number of layers
+        self.n_layers_ = len(layer_units)
+
+        # Output for regression
+        if not isinstance(self, ClassifierMixin):
+            self.out_activation_ = 'identity'
+        # Output for multi class
+        elif self.label_binarizer_.y_type_ == 'multiclass':
+            self.out_activation_ = 'softmax'
+        # Output for binary class and multi-label
+        else:
+            self.out_activation_ = 'logistic'
+            if self.loss == 'log_loss':
+                self.loss = 'binary_log_loss'
+
+        # Initialize coefficient and intercept layers
+        self.coefs_ = []
+        self.intercepts_ = []
+
+        for i in range(self.n_layers_ - 1):
+            rng = check_random_state(self.random_state)
+            coef_init, intercept_init = self._init_coef(layer_units[i],
+                                                        layer_units[i + 1],
+                                                        rng)
+            self.coefs_.append(coef_init)
+            self.intercepts_.append(intercept_init)
+
+        if self.algorithm in _STOCHASTIC_ALGOS:
+            self.loss_curve_ = []
+            self._no_improvement_count = 0
+            if self.early_stopping:
+                self.validation_scores_ = []
+                self.best_validation_score_ = -np.inf
+            else:
+                self.best_loss_ = np.inf
+
+    def _init_coef(self, fan_in, fan_out, rng):
+        if self.activation == 'logistic':
+            # Use the initialization method recommended by
+            # Glorot et al.
+            init_bound = np.sqrt(2. / (fan_in + fan_out))
+        elif self.activation == 'tanh':
+            init_bound = np.sqrt(6. / (fan_in + fan_out))
+        elif self.activation == 'relu':
+            # He et al instead maybe?
```

Comment:
No need for He et. al - as I understand this only helps in *super* deep networks e.g. 20+ layers. Orthogonal init might be useful but we can add it later if needed and might be overkill for the size of networks we will support